File: src/regen.cc
// Copyright 2016 Google Inc. All rights reserved
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// +build ignore

#include "regen.h"

#include <sys/stat.h>

#include <algorithm>
#include <future>
#include <memory>
#include <mutex>
#include <vector>

#include "affinity.h"
#include "fileutil.h"
#include "find.h"
#include "func.h"
#include "io.h"
#include "log.h"
#include "ninja.h"
#include "stats.h"
#include "strutil.h"

namespace {

#define RETURN_TRUE              \
  do {                           \
    if (g_flags.dump_kati_stamp) \
      needs_regen_ = true;       \
    else                         \
      return true;               \
  } while (0)

bool ShouldIgnoreDirty(std::string_view s) {
  Pattern pat(g_flags.ignore_dirty_pattern ? g_flags.ignore_dirty_pattern : "");
  Pattern nopat(
      g_flags.no_ignore_dirty_pattern ? g_flags.no_ignore_dirty_pattern : "");
  return pat.Match(s) && !nopat.Match(s);
}

class StampChecker {
  struct GlobResult {
    std::string pat;
    std::vector<std::string> result;
  };

  struct ShellResult {
    CommandOp op;
    std::string shell;
    std::string shellflag;
    std::string cmd;
    std::string result;
    std::vector<std::string> missing_dirs;
    std::vector<std::string> files;
    std::vector<std::string> read_dirs;
  };

 public:
  StampChecker() : needs_regen_(false) {}

  ~StampChecker() {
    for (GlobResult* gr : globs_) {
      delete gr;
    }
    for (ShellResult* sr : commands_) {
      delete sr;
    }
  }

  bool NeedsRegen(double start_time, const std::string& orig_args) {
    if (IsMissingOutputs())
      RETURN_TRUE;

    if (CheckStep1(orig_args))
      RETURN_TRUE;

    if (CheckStep2())
      RETURN_TRUE;

    if (!needs_regen_) {
      FILE* fp = fopen(GetNinjaStampFilename().c_str(), "rb+");
      if (!fp)
        return true;
      ScopedFile sfp(fp);
      if (fseek(fp, 0, SEEK_SET) < 0)
        PERROR("fseek");
      size_t r = fwrite(&start_time, sizeof(start_time), 1, fp);
      CHECK(r == 1);
    }
    return needs_regen_;
  }

 private:
  bool IsMissingOutputs() {
    if (!Exists(GetNinjaFilename())) {
      fprintf(stderr, "%s is missing, regenerating...\n",
              GetNinjaFilename().c_str());
      return true;
    }
    if (!Exists(GetNinjaShellScriptFilename())) {
      fprintf(stderr, "%s is missing, regenerating...\n",
              GetNinjaShellScriptFilename().c_str());
      return true;
    }
    return false;
  }

  bool CheckStep1(const std::string& orig_args) {
#define LOAD_INT(fp)                                               \
  ({                                                               \
    int v = LoadInt(fp);                                           \
    if (v < 0) {                                                   \
      fprintf(stderr, "incomplete kati_stamp, regenerating...\n"); \
      RETURN_TRUE;                                                 \
    }                                                              \
    v;                                                             \
  })

#define LOAD_STRING(fp, s)                                         \
  ({                                                               \
    if (!LoadString(fp, s)) {                                      \
      fprintf(stderr, "incomplete kati_stamp, regenerating...\n"); \
      RETURN_TRUE;                                                 \
    }                                                              \
  })

    const std::string& stamp_filename = GetNinjaStampFilename();
    FILE* fp = fopen(stamp_filename.c_str(), "rb");
    if (!fp) {
      if (g_flags.regen_debug)
        printf("%s: %s\n", stamp_filename.c_str(), strerror(errno));
      return true;
    }
    ScopedFile sfp(fp);

    double gen_time;
    size_t r = fread(&gen_time, sizeof(gen_time), 1, fp);
    gen_time_ = gen_time;
    if (r != 1) {
      fprintf(stderr, "incomplete kati_stamp, regenerating...\n");
      RETURN_TRUE;
    }
    if (g_flags.regen_debug)
      printf("Generated time: %f\n", gen_time);

    std::string s, s2;
    int num_files = LOAD_INT(fp);
    for (int i = 0; i < num_files; i++) {
      LOAD_STRING(fp, &s);
      double ts = GetTimestamp(s);
      if (gen_time < ts) {
        if (g_flags.regen_ignoring_kati_binary) {
          if (s == GetExecutablePath()) {
            fprintf(stderr, "%s was modified, ignored.\n", s.c_str());
            continue;
          }
        }
        if (ShouldIgnoreDirty(s)) {
          if (g_flags.regen_debug)
            printf("file %s: ignored (%f)\n", s.c_str(), ts);
          continue;
        }
        if (g_flags.dump_kati_stamp)
          printf("file %s: dirty (%f)\n", s.c_str(), ts);
        else
          fprintf(stderr, "%s was modified, regenerating...\n", s.c_str());
        RETURN_TRUE;
      } else if (g_flags.dump_kati_stamp) {
        printf("file %s: clean (%f)\n", s.c_str(), ts);
      }
    }

    int num_undefineds = LOAD_INT(fp);
    for (int i = 0; i < num_undefineds; i++) {
      LOAD_STRING(fp, &s);
      if (getenv(s.c_str())) {
        if (g_flags.dump_kati_stamp) {
          printf("env %s: dirty (unset => %s)\n", s.c_str(), getenv(s.c_str()));
        } else {
          fprintf(stderr, "Environment variable %s was set, regenerating...\n",
                  s.c_str());
        }
        RETURN_TRUE;
      } else if (g_flags.dump_kati_stamp) {
        printf("env %s: clean (unset)\n", s.c_str());
      }
    }

    int num_envs = LOAD_INT(fp);
    for (int i = 0; i < num_envs; i++) {
      LOAD_STRING(fp, &s);
      std::string_view val(getenv(s.c_str()));
      LOAD_STRING(fp, &s2);
      if (val != s2) {
        if (g_flags.dump_kati_stamp) {
          printf("env %s: dirty (%s => %.*s)\n", s.c_str(), s2.c_str(),
                 SPF(val));
        } else {
          fprintf(stderr,
                  "Environment variable %s was modified (%s => %.*s), "
                  "regenerating...\n",
                  s.c_str(), s2.c_str(), SPF(val));
        }
        RETURN_TRUE;
      } else if (g_flags.dump_kati_stamp) {
        printf("env %s: clean (%.*s)\n", s.c_str(), SPF(val));
      }
    }

    int num_globs = LOAD_INT(fp);
    std::string pat;
    for (int i = 0; i < num_globs; i++) {
      GlobResult* gr = new GlobResult;
      globs_.push_back(gr);

      LOAD_STRING(fp, &gr->pat);
      int num_files = LOAD_INT(fp);
      gr->result.resize(num_files);
      for (int j = 0; j < num_files; j++) {
        LOAD_STRING(fp, &gr->result[j]);
      }
    }

    int num_crs = LOAD_INT(fp);
    for (int i = 0; i < num_crs; i++) {
      ShellResult* sr = new ShellResult;
      commands_.push_back(sr);
      sr->op = static_cast<CommandOp>(LOAD_INT(fp));
      LOAD_STRING(fp, &sr->shell);
      LOAD_STRING(fp, &sr->shellflag);
      LOAD_STRING(fp, &sr->cmd);
      LOAD_STRING(fp, &sr->result);

      std::string file;
      // Ignore debug info
      LOAD_STRING(fp, &file);
      LOAD_INT(fp);

      if (sr->op == CommandOp::FIND) {
        int num_missing_dirs = LOAD_INT(fp);
        for (int j = 0; j < num_missing_dirs; j++) {
          LOAD_STRING(fp, &s);
          sr->missing_dirs.push_back(s);
        }
        int num_files = LOAD_INT(fp);
        for (int j = 0; j < num_files; j++) {
          LOAD_STRING(fp, &s);
          sr->files.push_back(s);
        }
        int num_read_dirs = LOAD_INT(fp);
        for (int j = 0; j < num_read_dirs; j++) {
          LOAD_STRING(fp, &s);
          sr->read_dirs.push_back(s);
        }
      }
    }

    LoadString(fp, &s);
    if (orig_args != s) {
      fprintf(stderr, "arguments changed, regenerating...\n");
      RETURN_TRUE;
    }

    return needs_regen_;
  }

  bool CheckGlobResult(const GlobResult* gr, std::string* err) {
    COLLECT_STATS("glob time (regen)");
    const auto& files = Glob(gr->pat.c_str());
    bool needs_regen = files.size() != gr->result.size();
    for (size_t i = 0; i < gr->result.size(); i++) {
      if (!needs_regen) {
        if (files[i] != gr->result[i]) {
          needs_regen = true;
          break;
        }
      }
    }
    if (needs_regen) {
      if (ShouldIgnoreDirty(gr->pat)) {
        if (g_flags.dump_kati_stamp) {
          printf("wildcard %s: ignored\n", gr->pat.c_str());
        }
        return false;
      }
      if (g_flags.dump_kati_stamp) {
        printf("wildcard %s: dirty\n", gr->pat.c_str());
      } else {
        *err = StringPrintf("wildcard(%s) was changed, regenerating...\n",
                            gr->pat.c_str());
      }
    } else if (g_flags.dump_kati_stamp) {
      printf("wildcard %s: clean\n", gr->pat.c_str());
    }
    return needs_regen;
  }

  bool ShouldRunCommand(const ShellResult* sr) {
    if (sr->op != CommandOp::FIND)
      return true;

    COLLECT_STATS("stat time (regen)");
    for (const std::string& dir : sr->missing_dirs) {
      if (Exists(dir))
        return true;
    }
    for (const std::string& file : sr->files) {
      if (!Exists(file))
        return true;
    }
    for (const std::string& dir : sr->read_dirs) {
      // We assume we rarely do a significant change for the top
      // directory which affects the results of find command.
      if (dir == "" || dir == "." || ShouldIgnoreDirty(dir))
        continue;

      struct stat st;
      if (lstat(dir.c_str(), &st) != 0) {
        return true;
      }
      double ts = GetTimestampFromStat(st);
      if (gen_time_ < ts) {
        return true;
      }
      if (S_ISLNK(st.st_mode)) {
        ts = GetTimestamp(dir);
        if (ts < 0 || gen_time_ < ts)
          return true;
      }
    }
    return false;
  }

  bool CheckShellResult(const ShellResult* sr, std::string* err) {
    if (sr->op == CommandOp::READ_MISSING) {
      if (Exists(sr->cmd)) {
        if (g_flags.dump_kati_stamp)
          printf("file %s: dirty\n", sr->cmd.c_str());
        else
          *err = StringPrintf("$(file <%s) was changed, regenerating...\n",
                              sr->cmd.c_str());
        return true;
      }
      if (g_flags.dump_kati_stamp)
        printf("file %s: clean\n", sr->cmd.c_str());
      return false;
    }

    if (sr->op == CommandOp::READ) {
      double ts = GetTimestamp(sr->cmd);
      if (gen_time_ < ts) {
        if (g_flags.dump_kati_stamp)
          printf("file %s: dirty\n", sr->cmd.c_str());
        else
          *err = StringPrintf("$(file <%s) was changed, regenerating...\n",
                              sr->cmd.c_str());
        return true;
      }
      if (g_flags.dump_kati_stamp)
        printf("file %s: clean\n", sr->cmd.c_str());
      return false;
    }

    if (sr->op == CommandOp::WRITE || sr->op == CommandOp::APPEND) {
      FILE* f =
          fopen(sr->cmd.c_str(), (sr->op == CommandOp::WRITE) ? "wb" : "ab");
      if (f == NULL) {
        PERROR("fopen");
      }

      if (fwrite(&sr->result[0], sr->result.size(), 1, f) != 1) {
        PERROR("fwrite");
      }

      if (fclose(f) != 0) {
        PERROR("fclose");
      }

      if (g_flags.dump_kati_stamp)
        printf("file %s: clean (write)\n", sr->cmd.c_str());
      return false;
    }

    if (!ShouldRunCommand(sr)) {
      if (g_flags.regen_debug)
        printf("shell %s: clean (no rerun)\n", sr->cmd.c_str());
      return false;
    }

    FindCommand fc;
    if (fc.Parse(sr->cmd) && !fc.chdir.empty() && ShouldIgnoreDirty(fc.chdir)) {
      if (g_flags.dump_kati_stamp)
        printf("shell %s: ignored\n", sr->cmd.c_str());
      return false;
    }

    COLLECT_STATS_WITH_SLOW_REPORT("shell time (regen)", sr->cmd.c_str());
    std::string result;
    RunCommand(sr->shell, sr->shellflag, sr->cmd, RedirectStderr::DEV_NULL,
               &result);
    FormatForCommandSubstitution(&result);
    if (sr->result != result) {
      if (g_flags.dump_kati_stamp) {
        printf("shell %s: dirty\n", sr->cmd.c_str());
      } else {
        *err = StringPrintf("$(shell %s) was changed, regenerating...\n",
                            sr->cmd.c_str());
        //*err += StringPrintf("%s => %s\n", expected.c_str(), result.c_str());
      }
      return true;
    } else if (g_flags.regen_debug) {
      printf("shell %s: clean (rerun)\n", sr->cmd.c_str());
    }
    return false;
  }

  bool CheckStep2() {
    auto glob_future = std::async([this]() {
      std::string err;
      // TODO: Make glob cache thread safe and create a task for each glob.
      SetAffinityForSingleThread();
      for (GlobResult* gr : globs_) {
        if (CheckGlobResult(gr, &err)) {
          std::unique_lock<std::mutex> lock(mu_);
          if (!needs_regen_) {
            needs_regen_ = true;
            msg_ = err;
          }
          break;
        }
      }
    });

    auto shell_future = std::async([this]() {
      SetAffinityForSingleThread();
      for (ShellResult* sr : commands_) {
        std::string err;
        if (CheckShellResult(sr, &err)) {
          std::unique_lock<std::mutex> lock(mu_);
          if (!needs_regen_) {
            needs_regen_ = true;
            msg_ = err;
          }
        }
      }
    });

    glob_future.wait();
    shell_future.wait();
    if (needs_regen_) {
      fprintf(stderr, "%s", msg_.c_str());
    }
    return needs_regen_;
  }

 private:
  double gen_time_;
  std::vector<GlobResult*> globs_;
  std::vector<ShellResult*> commands_;
  std::mutex mu_;
  bool needs_regen_;
  std::string msg_;
};

}  // namespace

bool NeedsRegen(double start_time, const std::string& orig_args) {
  return StampChecker().NeedsRegen(start_time, orig_args);
}

File: src-rs/regen.rs
/*
Copyright 2025 Google LLC

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use crate::{
    collect_stats, collect_stats_with_slow_report,
    fileutil::{glob, run_command},
    flags::FLAGS,
    func::CommandOp,
    io::{dump_systemtime, load_int, load_string, load_systemtime, load_usize, load_vec_string},
    ninja::{get_ninja_filename, get_ninja_shell_script_filename, get_ninja_stamp_filename},
    strutil::format_for_command_substitution,
};
use anyhow::Result;
use bytes::Bytes;
use parking_lot::Mutex;
use std::{
    ffi::{OsStr, OsString},
    fs::OpenOptions,
    io::{BufReader, Write},
    os::unix::ffi::{OsStrExt, OsStringExt},
    time::SystemTime,
};

fn should_ignore_dirty(s: &[u8]) -> bool {
    let Some(ignore) = &FLAGS.ignore_dirty_pattern else {
        return false;
    };
    ignore.matches(s)
        && !FLAGS
            .no_ignore_dirty_pattern
            .as_ref()
            .map(|p| p.matches(s))
            .unwrap_or(false)
}

struct GlobResult {
    pat: Bytes,
    result: Vec<Bytes>,
}

struct ShellResult {
    op: CommandOp,
    shell: OsString,
    shellflag: OsString,
    cmd: OsString,
    result: Vec<u8>,
    missing_dirs: Vec<OsString>,
    files: Vec<OsString>,
    read_dirs: Vec<OsString>,
}

struct StampChecker {
    gen_time: Option<SystemTime>,
    globs: Vec<GlobResult>,
    commands: Vec<ShellResult>,
    needs_regen: bool,
}

macro_rules! load {
    ($v:expr) => {
        match $v {
            Some(s) => s,
            None => {
                eprintln!("incomplete kati_stamp, regenerating...");
                return true;
            }
        }
    };
}

impl StampChecker {
    fn new() -> Self {
        Self {
            gen_time: None,
            globs: Vec::new(),
            commands: Vec::new(),
            needs_regen: false,
        }
    }

    fn needs_regen(&mut self, start_time: SystemTime, orig_args: &OsStr) -> bool {
        if Self::is_missing_outputs() {
            return true;
        }

        if self.check_step1(orig_args) {
            return true;
        }

        if self.check_step2().unwrap_or(true) {
            return true;
        }

        if !self.needs_regen {
            let mut opts = OpenOptions::new();
            let Ok(mut fp) = opts.write(true).open(get_ninja_stamp_filename()) else {
                return true;
            };
            dump_systemtime(&mut fp, &start_time).unwrap();
        }
        self.needs_regen
    }

    fn is_missing_outputs() -> bool {
        let f = get_ninja_filename();
        if !std::fs::exists(&f).is_ok_and(|b| b) {
            eprintln!("{} is missing, regenerating...", f.to_string_lossy());
            return true;
        }
        let f = get_ninja_shell_script_filename();
        if !std::fs::exists(&f).is_ok_and(|b| b) {
            eprintln!("{} is missing, regenerating...", f.to_string_lossy());
            return true;
        }
        false
    }

    fn check_step1(&mut self, orig_args: &OsStr) -> bool {
        let stamp_filename = get_ninja_stamp_filename();
        let fp = match std::fs::File::open(&stamp_filename) {
            Ok(fp) => fp,
            Err(err) => {
                if FLAGS.regen_debug {
                    println!("{stamp_filename:?}: {err}")
                }
                return true;
            }
        };
        let mut fp = BufReader::new(fp);
        let fp = &mut fp;

        let gen_time = load!(load_systemtime(fp));
        self.gen_time = Some(gen_time);
        if FLAGS.regen_debug {
            println!("Generated time: {:?}", self.gen_time);
        }

        let files = load!(load_vec_string(fp));
        for s in files {
            let ts = std::fs::metadata(&s).and_then(|m| m.modified());
            if ts.as_ref().is_ok_and(|ts| gen_time >= *ts) {
                if FLAGS.dump_kati_stamp {
                    println!("file {s:?}: clean ({:?})", ts.unwrap())
                }
            } else {
                if FLAGS.regen_ignoring_kati_binary && s == std::env::current_exe().unwrap() {
                    eprintln!("{s:?} was modified, ignored.");
                    continue;
                }
                if should_ignore_dirty(s.as_bytes()) {
                    if FLAGS.regen_debug {
                        println!("file {s:?}: ignored ({:?})", ts.unwrap());
                    }
                    continue;
                }
                if FLAGS.dump_kati_stamp {
                    println!("file {s:?}: dirty ({:?})", ts.unwrap());
                } else {
                    eprintln!("{} was modified, regenerating...", s.to_string_lossy());
                }
                return true;
            }
        }

        let undefineds = load!(load_vec_string(fp));
        for s in undefineds {
            if let Ok(v) = std::env::var(&s) {
                if FLAGS.dump_kati_stamp {
                    println!("env {s:?}: dirty (unset => {v:?})");
                } else {
                    eprintln!(
                        "Environment variable {} was set, regenerating...",
                        s.to_string_lossy()
                    );
                }
                return true;
            } else if FLAGS.dump_kati_stamp {
                println!("env {s:?}: clean (unset)");
            }
        }

        let num_envs = load!(load_usize(fp));
        for _ in 0..num_envs {
            let s = load!(load_string(fp));
            let val = std::env::var_os(&s).unwrap_or_default();
            let s2 = load!(load_string(fp));
            if val != s2 {
                if FLAGS.dump_kati_stamp {
                    println!("env {s:?}: dirty ({s2:?} => {val:?})")
                } else {
                    eprintln!(
                        "Environment variable {} was modified ({:?} => {:?}), regenerating...",
                        s.to_string_lossy(),
                        s2.to_string_lossy(),
                        val.to_string_lossy(),
                    );
                }
                return true;
            } else if FLAGS.dump_kati_stamp {
                println!("env {s:?}: clean ({val:?})")
            }
        }

        let num_globs = load!(load_usize(fp));
        for _ in 0..num_globs {
            let pat = load!(load_string(fp));
            let result = load!(load_vec_string(fp));
            self.globs.push(GlobResult {
                pat: Bytes::from(pat.into_vec()),
                result: result
                    .into_iter()
                    .map(|r| Bytes::from(r.into_vec()))
                    .collect(),
            })
        }

        let num_crs = load!(load_usize(fp));
        for _ in 0..num_crs {
            let op = load!(load_int(fp).and_then(CommandOp::from_int));
            let shell = load!(load_string(fp));
            let shellflag = load!(load_string(fp));
            let cmd = load!(load_string(fp));
            let result = load!(load_string(fp));
            let mut sr = ShellResult {
                op,
                shell,
                shellflag,
                cmd,
                result: result.into_vec(),
                missing_dirs: Vec::new(),
                files: Vec::new(),
                read_dirs: Vec::new(),
            };

            // Ignore debug info
            load!(load_string(fp));
            load!(load_int(fp));

            if op == CommandOp::Find {
                sr.missing_dirs = load!(load_vec_string(fp));
                sr.files = load!(load_vec_string(fp));
                sr.read_dirs = load!(load_vec_string(fp));
            }
            self.commands.push(sr);
        }

        let s = load!(load_string(fp));
        if orig_args != s {
            eprintln!("arguments changed, regenerating...");
            return true;
        }

        self.needs_regen
    }

    fn check_glob_result(gr: &GlobResult, err: &mut String) -> bool {
        collect_stats!("glob time (regen)");
        let files = glob(gr.pat.clone());
        let needs_regen = if let Ok(files) = files.as_ref() {
            files != &gr.result
        } else {
            true
        };
        if needs_regen {
            if should_ignore_dirty(&gr.pat) {
                if FLAGS.dump_kati_stamp {
                    println!("wildcard {:?}: ignored", gr.pat);
                }
                return false;
            }
            if FLAGS.dump_kati_stamp {
                println!("wildcard {:?}: dirty", gr.pat);
            } else {
                *err = format!(
                    "wildcard({}) was changed, regenerating...",
                    String::from_utf8_lossy(&gr.pat)
                );
            }
        } else if FLAGS.dump_kati_stamp {
            println!("wildcard {:?}: clean", gr.pat);
        }
        needs_regen
    }

    fn should_run_command(sr: &ShellResult, gen_time: SystemTime) -> bool {
        if sr.op != CommandOp::Find {
            return true;
        }

        collect_stats!("stat time (regen)");
        for dir in &sr.missing_dirs {
            if std::fs::exists(dir).unwrap_or(false) {
                return true;
            }
        }
        for file in &sr.files {
            if !std::fs::exists(file).unwrap_or(false) {
                return true;
            }
        }
        for dir in &sr.read_dirs {
            // We assume we rarely do a significant change for the top
            // directory which affects the results of find command.
            if dir.is_empty() || dir == "." || should_ignore_dirty(dir.as_bytes()) {
                continue;
            }

            let Ok(md) = std::fs::symlink_metadata(dir) else {
                return true;
            };
            let Ok(ts) = md.modified() else {
                return true;
            };
            if gen_time < ts {
                return true;
            }
            if md.is_symlink() {
                let Ok(ts) = std::fs::metadata(dir).and_then(|md| md.modified()) else {
                    return true;
                };
                if gen_time < ts {
                    return true;
                }
            }
        }
        false
    }

    fn check_shell_result(
        sr: &ShellResult,
        gen_time: SystemTime,
        err: &mut String,
    ) -> Result<bool> {
        if sr.op == CommandOp::ReadMissing {
            if std::fs::exists(&sr.cmd).unwrap_or(false) {
                if FLAGS.dump_kati_stamp {
                    println!("file {:?}: dirty", sr.cmd);
                } else {
                    *err = format!(
                        "$(file <{}) was changed, regenerating...",
                        sr.cmd.to_string_lossy()
                    );
                }
                return Ok(true);
            }
            if FLAGS.dump_kati_stamp {
                println!("file {:?}: clean", sr.cmd);
            }
            return Ok(false);
        }

        if sr.op == CommandOp::Read {
            let ts = std::fs::metadata(&sr.cmd).and_then(|md| md.modified());
            if ts.is_ok_and(|ts| gen_time < ts) {
                if FLAGS.dump_kati_stamp {
                    println!("file {:?}: dirty", sr.cmd);
                } else {
                    *err = format!(
                        "$(file <{}) was changed, regenerating...",
                        sr.cmd.to_string_lossy()
                    );
                }
                return Ok(true);
            }
            if FLAGS.dump_kati_stamp {
                println!("file {:?}: clean", sr.cmd);
            }
            return Ok(false);
        }

        if sr.op == CommandOp::Write || sr.op == CommandOp::Append {
            let mut opts = OpenOptions::new();
            opts.write(true)
                .create(true)
                .append(sr.op == CommandOp::Append);
            let mut f = opts.open(&sr.cmd)?;
            f.write_all(&sr.result)?;

            if FLAGS.dump_kati_stamp {
                println!("file {:?}: clean (write)", sr.cmd);
            }
            return Ok(false);
        }

        if !Self::should_run_command(sr, gen_time) {
            if FLAGS.regen_debug {
                println!("shell {:?}: clean (no rerun)", sr.cmd);
            }
            return Ok(false);
        }

        let cmd = Bytes::from(sr.cmd.clone().into_vec());
        if let Some(fc) = crate::find::parse(&cmd)?
            && fc.chdir.is_some_and(|d| should_ignore_dirty(&d))
        {
            if FLAGS.dump_kati_stamp {
                println!("shell {:?}: ignored", sr.cmd);
            }
            return Ok(false);
        }

        collect_stats_with_slow_report!("shell time (regen)", &sr.cmd);
        let (_status, output) = run_command(
            sr.shell.as_bytes(),
            sr.shellflag.as_bytes(),
            &cmd,
            crate::fileutil::RedirectStderr::DevNull,
        )?;
        let output = format_for_command_substitution(output);
        if sr.result != output {
            if FLAGS.dump_kati_stamp {
                println!("shell {:?}: dirty", sr.cmd);
            } else {
                *err = format!(
                    "$(shell {}) was changed, regenerating...",
                    sr.cmd.to_string_lossy()
                );
            }
            return Ok(true);
        } else if FLAGS.regen_debug {
            println!("shell {:?}: clean (rerun)", sr.cmd);
        }
        Ok(false)
    }

    fn check_step2(&mut self) -> Result<bool> {
        let needs_regen: Mutex<Result<bool>> = Mutex::new(Ok(false));

        std::thread::scope(|s| {
            s.spawn(|| {
                let mut err = String::new();
                for gr in &self.globs {
                    if Self::check_glob_result(gr, &mut err) {
                        let mut needs_regen = needs_regen.lock();
                        if let Ok(false) = *needs_regen {
                            *needs_regen = Ok(true);
                            eprintln!("{err}");
                        }
                        break;
                    }
                }
            });
            s.spawn(|| {
                let mut err = String::new();
                for sr in &self.commands {
                    match Self::check_shell_result(sr, self.gen_time.unwrap(), &mut err) {
                        Ok(true) => {
                            let mut needs_regen = needs_regen.lock();
                            if let Ok(false) = *needs_regen {
                                *needs_regen = Ok(true);
                                eprintln!("{err}");
                            }
                            break;
                        }
                        Ok(false) => {}
                        Err(e) => {
                            let mut needs_regen = needs_regen.lock();
                            if let Ok(false) = *needs_regen {
                                *needs_regen = Err(e);
                            }
                            break;
                        }
                    }
                }
            });
        });

        needs_regen.into_inner()
    }
}

pub fn needs_regen(start_time: SystemTime, orig_args: &OsStr) -> bool {
    StampChecker::new().needs_regen(start_time, orig_args)
}
